lib/matplotlib/_mathtext.py

```diff
@@ -1874,15 +1865,14 @@ def __init__(self):
                | Error("Expected a delimiter"))
         )
 
-        p.math          <<= OneOrMore(p.token)
-
-        p.math_string   <<= QuotedString('$', '\\', unquoteResults=False)
-
-        p.non_math      <<= Regex(r"(?:(?:\\[$])|[^$])*").leaveWhitespace()
-
-        p.main          <<= (
+        # Leaf definitions.
+        p.math          = OneOrMore(p.token)
+        p.math_string   = QuotedString('$', '\\', unquoteResults=False)
+        p.non_math      = Regex(r"(?:(?:\\[$])|[^$])*").leaveWhitespace()
+        p.main          = (
             p.non_math + ZeroOrMore(p.math_string + p.non_math) + StringEnd()
         )
+        set_names_and_parse_actions()  # for leaf definitions.
```

Comment:
Why do we need to call this 3x?

Reply:
This is because e.g. `float_literal` is defined at the very top (root definition), but then gets used as `float_literal("rulesize")` in the mutually recursive definitions block; this creates a *copy* of `float_literal` and adds a name ("rulesize") to it.  Therefore, `float_literal` must *already* have its parse action set at that point.  So it's not possible to set all parse actions at the end after all definitions.  (It's also not possible to set the parse actions all together after declaring all blocks (as was done before), as the point of the PR is exactly to not have to declare blocks anymore.)

Technically, this does mean that the same parse action gets re-set three times on the root elements and two times on the mutually recursive elements, but this isn't a problem at all, afaict.